x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -112,104 +130,130 @@ public ModelLoadingService(TrainedModelProvider trainedModelProvider,
+     * @param modelActionListener the listener to alert when the model has been retrieved
+     */
+    public void getModelForSearch(String modelId, ActionListener<Model> modelActionListener) {
+        getModel(modelId, Consumer.SEARCH, modelActionListener);
+    }
+
+    /**
+     * Gets the model referenced by `modelId` and responds to the listener.
+     * <p>
      * This method first checks the local LRU cache for the model. If it is present, it is returned from cache.
+     * <p>
+     * In the case of search if the model is not present one of the following occurs:
+     * - If it is currently being loaded the `modelActionListener`
+     * is added to the list of listeners to be alerted when the model is fully loaded.
+     * - Otherwise the model is loaded and cached
      *
-     * If it is not present, one of the following occurs:
+     * In the case of an ingest processor if it is not present, one of the following occurs:
+     * <p>
+     * - If the model is referenced by a pipeline and is currently being loaded, the `modelActionListener`
+     * is added to the list of listeners to be alerted when the model is fully loaded.
+     * - If the model is referenced by a pipeline and is currently NOT being loaded, a new load attempt is made and the resulting
+     * model will attempt to be cached for future reference
+     * - If the models is NOT referenced by a pipeline, the model is simply loaded from the index and given to the listener.
+     * It is not cached.
      *
-     *  - If the model is referenced by a pipeline and is currently being loaded, the `modelActionListener`
-     *    is added to the list of listeners to be alerted when the model is fully loaded.
-     *  - If the model is referenced by a pipeline and is currently NOT being loaded, a new load attempt is made and the resulting
-     *    model will attempt to be cached for future reference
-     *  - If the models is NOT referenced by a pipeline, the model is simply loaded from the index and given to the listener.
-     *    It is not cached.
+     * The main difference being that models for search are always cached whereas pipeline models
+     * are only cached if they are referenced by an ingest pipeline
      *
-     * @param modelId the model to get
+     * @param modelId             the model to get
+     * @param consumer            which feature is requesting the model
      * @param modelActionListener the listener to alert when the model has been retrieved.
      */
-    public void getModel(String modelId, ActionListener<Model> modelActionListener) {
-        LocalModel cachedModel = localModelCache.get(modelId);
+    private void getModel(String modelId, Consumer consumer, ActionListener<Model> modelActionListener) {
+        ModelAndConsumer cachedModel = localModelCache.get(modelId);
         if (cachedModel != null) {
-            modelActionListener.onResponse(cachedModel);
+            cachedModel.consumers.add(consumer);
```

Comment:
There is a pretty frustrating race condition here, unsure if it is worth fixing.

- `ModelAndConsumer cachedModel = localModelCache.get(modelId);` returns from cache successfully
- Model is removed from cache as it is only referenced by a pipeline (updated via cluster state change)
- Search adds itself as a consume too late and it is evicted

The referenced model from the `get` call will still be around. But the next call for the same model will load from the index again :/.

The only way I could see of fixing this is locking here (which defeats the purpose of the fast `get`). Since the cost is low, it might just be OK to have this as it is recoverable later.

Reply:
Good point it is recoverable and there will always be cases where the model is requested after it has been evicted so it is not too bad. But I also think if the cache entry is being modified we need to ensure that change is visible to other threads and synchronised is an easy way of doing that. 

I can't decide which to do here 🤔

Reply:
>  But I also think if the cache entry is being modified we need to ensure that change is visible to other threads and synchronised is an easy way of doing that.

Are enum set changes visible to other threads? If they are not, we should use a MT safe/acceptable set for storing which consumers have the model.

I don't think we should synchronize here. Simply having something `volatile` somewhere in the call path will work.

Reply:
The enum set is a nice fast efficient bit set under the hood I'd rather not change that for something thread safe. 

It is the eviction where we really care about reading the changes to the enum set as that is where the check for who is using the model, that code is synchronised so no visibility problems there.

I think we are good with the fast unsynchronised lookup, it is optimised for the common case and we are eventually consistent anyway

Reply:
> I think we are good with the fast unsynchronised lookup, it is optimised for the common case and we are eventually consistent anyway

OK, sounds good. The PR looks like that has been removed?

Reply:
Oops, reinstated. I'll put that down to a fat fingered rebase